metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py

```diff
@@ -3457,33 +3493,88 @@ def _create_workbook_properties(
         return {"permissions": json.dumps(groups)} if len(groups) > 0 else None
 
     def ingest_tableau_site(self):
+        self.report.report_ingestion_stage_start(
+            f"Ingesting Tableau Site: {self.site_id} {self.site_content_url}"
+        )
+
         # Initialise the dictionary to later look-up for chart and dashboard stat
         if self.config.extract_usage_stats:
-            self._populate_usage_stat_registry()
+            with PerfTimer() as timer:
+                self._populate_usage_stat_registry()
+                self.report.extract_usage_stats_timer[self.site_id] = round(
+                    timer.elapsed_seconds(), 2
+                )
 
         if self.config.permission_ingestion:
-            self._fetch_groups()
+            with PerfTimer() as timer:
+                self._fetch_groups()
+                self.report.fetch_groups_timer[self.site_id] = round(
+                    timer.elapsed_seconds(), 2
+                )
 
         # Populate the map of database names and database hostnames to be used later to map
         # databases to platform instances.
         if self.config.database_hostname_to_platform_instance_map:
-            self._populate_database_server_hostname_map()
+            with PerfTimer() as timer:
+                self._populate_database_server_hostname_map()
+                self.report.populate_database_server_hostname_map_timer[
+                    self.site_id
+                ] = round(timer.elapsed_seconds(), 2)
 
-        self._populate_projects_registry()
+        with PerfTimer() as timer:
+            self._populate_projects_registry()
+            self.report.populate_projects_registry_timer[self.site_id] = round(
+                timer.elapsed_seconds(), 2
+            )
 
         if self.config.add_site_container:
             yield from self.emit_site_container()
         yield from self.emit_project_containers()
-        yield from self.emit_workbooks()
+
+        with PerfTimer() as timer:
+            yield from self.emit_workbooks()
+            self.report.emit_workbooks_timer[self.site_id] = round(
+                timer.elapsed_seconds(), 2
+            )
+
         if self.sheet_ids:
-            yield from self.emit_sheets()
+            with PerfTimer() as timer:
+                yield from self.emit_sheets()
+                self.report.emit_sheets_timer[self.site_id] = round(
+                    timer.elapsed_seconds(), 2
+                )
+
         if self.dashboard_ids:
-            yield from self.emit_dashboards()
+            with PerfTimer() as timer:
+                yield from self.emit_dashboards()
+                self.report.emit_dashboards_timer[self.site_id] = round(
```

Comment:
PerfTimer objects can go directly into the report, and will get automatically formatted nicely - it implements `as_obj`, which is used here https://github.com/datahub-project/datahub/blob/2e14f70864ef3010cb1893445d8967ff59b296b0/metadata-ingestion/src/datahub/ingestion/api/report.py#L41

so we could do something like `with self.report.emit_dashboards_timer.setdefault(self.site_id, PerfTimer()):`

Reply:
although we might need to make some tweaks to TopKDict to make that work, so may not be worth it